File: fabric/src/main/java/net/shiroha233/roadweaver/network/fabric/MapNetworkFabric.java
package net.shiroha233.roadweaver.network.fabric;

import net.fabricmc.fabric.api.client.networking.v1.ClientPlayNetworking;
import net.fabricmc.fabric.api.networking.v1.PayloadTypeRegistry;
import net.fabricmc.fabric.api.networking.v1.ServerPlayNetworking;
import net.minecraft.core.BlockPos;
import net.minecraft.network.RegistryFriendlyByteBuf;
import net.minecraft.network.codec.StreamCodec;
import net.minecraft.network.chat.Component;
import net.minecraft.network.protocol.common.custom.CustomPacketPayload;
import net.minecraft.resources.Identifier;
import net.minecraft.server.level.ServerPlayer;
import net.minecraft.world.level.levelgen.Heightmap;
import net.shiroha233.roadweaver.client.map.data.MapDataCollector;
import net.shiroha233.roadweaver.client.map.data.MapSnapshot;
import net.shiroha233.roadweaver.client26.RoadMapScreen26;
import net.shiroha233.roadweaver.config.ConfigService;
import net.shiroha233.roadweaver.config.ModConfig;
import net.shiroha233.roadweaver.core.model.ConnectionStatus;
import net.shiroha233.roadweaver.core.model.StructureConnection;
import net.shiroha233.roadweaver.helpers.PermissionCompat;
import net.shiroha233.roadweaver.map.permission.MapAccessService;
import net.shiroha233.roadweaver.network.MapSnapshotCodec;
import net.shiroha233.roadweaver.persistence.WorldDataProvider;
import net.shiroha233.roadweaver.runtime.ThreadPoolManager;

import java.util.ArrayList;
import java.util.List;
import java.util.concurrent.CompletableFuture;

/** Fabric networking implementation for the native Minecraft 26.2 RoadWeaver map UI. */
public final class MapNetworkFabric {
    private static boolean payloadTypesRegistered;

    private MapNetworkFabric() {}

    private static synchronized void registerPayloadTypes() {
        if (payloadTypesRegistered) return;
        PayloadTypeRegistry.serverboundPlay().register(RequestRectPayload.TYPE, RequestRectPayload.CODEC);
        PayloadTypeRegistry.serverboundPlay().register(TeleportRequestPayload.TYPE, TeleportRequestPayload.CODEC);
        PayloadTypeRegistry.serverboundPlay().register(ManualConnectPayload.TYPE, ManualConnectPayload.CODEC);
        PayloadTypeRegistry.clientboundPlay().register(SnapshotPayload.TYPE, SnapshotPayload.CODEC);
        PayloadTypeRegistry.clientboundPlay().register(TeleportAckPayload.TYPE, TeleportAckPayload.CODEC);
        PayloadTypeRegistry.clientboundPlay().register(AccessSyncPayload.TYPE, AccessSyncPayload.CODEC);
        payloadTypesRegistered = true;
    }

    public static void registerServerReceivers() {
        registerPayloadTypes();

        ServerPlayNetworking.registerGlobalReceiver(RequestRectPayload.TYPE, (payload, context) -> {
            ServerPlayer sp = context.player();
            if (!MapAccessService.canOpenMap(sp)) {
                context.server().execute(() -> syncMapAccess(sp));
                return;
            }
            int cx = (int)Math.round(sp.getX());
            int cz = (int)Math.round(sp.getZ());
            int computedRadiusBlocks;
            try {
                ModConfig cfg = ConfigService.get();
                if (cfg.highway().enabled()) computedRadiusBlocks = Math.max(16, cfg.highway().planningRadiusBlocks());
                else {
                    int radiusChunks = cfg.planning().dynamicPlanEnabled()
                            ? cfg.planning().dynamicPlanRadiusChunks()
                            : cfg.planning().initialPlanRadiusChunks();
                    computedRadiusBlocks = Math.max(1, radiusChunks) * 16;
                }
            } catch (Throwable ignored) {
                computedRadiusBlocks = 4096;
            }
            final int radiusBlocks = Math.max(16, computedRadiusBlocks);

            CompletableFuture.supplyAsync(() -> {
                var level = sp.level();
                Identifier actualDimensionId = level.dimension().identifier();
                MapSnapshot snapshot = MapDataCollector.build(level, payload.minX(), payload.minZ(), payload.maxX(), payload.maxZ(), cx, cz, radiusBlocks);
                return new SnapshotPayload(payload.requestSeq(), actualDimensionId, snapshot);
            }, ThreadPoolManager.computeExecutor()).thenAccept(reply -> context.server().execute(() -> {
                if (!sp.isRemoved() && !sp.hasDisconnected()) ServerPlayNetworking.send(sp, reply);
            }));
        });

        ServerPlayNetworking.registerGlobalReceiver(TeleportRequestPayload.TYPE, (payload, context) -> context.server().execute(() -> {
            ServerPlayer sp = context.player();
            if (!(sp.isCreative() || PermissionCompat.hasCommandLevel2(sp))) {
                ServerPlayNetworking.send(sp, new TeleportAckPayload(false, 0, 0, 0));
                return;
            }
            int x = payload.x();
            int z = payload.z();
            var level = sp.level();
            level.getChunk(x >> 4, z >> 4);
            int y = level.getHeight(Heightmap.Types.MOTION_BLOCKING_NO_LEAVES, x, z);
            if (y <= level.getMinY()) y = level.getSeaLevel() + 1; else y += 1;
            sp.teleportTo(level, x + 0.5D, y, z + 0.5D, java.util.Set.of(), sp.getYRot(), sp.getXRot(), false);
            ServerPlayNetworking.send(sp, new TeleportAckPayload(true, x, y, z));
        }));

        ServerPlayNetworking.registerGlobalReceiver(ManualConnectPayload.TYPE, (payload, context) -> context.server().execute(() -> {
            ServerPlayer sp = context.player();
            if (!PermissionCompat.hasCommandLevel2(sp)) return;
            var level = sp.level();
            WorldDataProvider provider = WorldDataProvider.getInstance();
            List<StructureConnection> origin = provider.getStructureConnections(level);
            List<StructureConnection> list = origin != null ? new ArrayList<>(origin) : new ArrayList<>();
            BlockPos a = new BlockPos(payload.ax(), 0, payload.az());
            BlockPos b = new BlockPos(payload.bx(), 0, payload.bz());
            boolean exists = list.stream().anyMatch(c -> (c.from().equals(a) && c.to().equals(b)) || (c.from().equals(b) && c.to().equals(a)));
            if (!exists) {
                list.add(new StructureConnection(a, b, ConnectionStatus.PLANNED));
                provider.setStructureConnections(level, list);
            }
        }));
    }

    public static void registerClientReceivers() {
        registerPayloadTypes();
        ClientPlayNetworking.registerGlobalReceiver(SnapshotPayload.TYPE, (payload, context) -> context.client().execute(() -> {
            if (context.client().gui.screen() instanceof RoadMapScreen26 screen) {
                screen.acceptSnapshot(payload.requestSeq(), payload.dimensionId(), payload.snapshot());
            }
        }));
        ClientPlayNetworking.registerGlobalReceiver(TeleportAckPayload.TYPE, (payload, context) -> context.client().execute(() -> {
            Component message = payload.ok()
                    ? Component.translatable("gui.roadweaver.map.teleport.success_pos", payload.x(), payload.y(), payload.z())
                    : Component.translatable("gui.roadweaver.map.teleport.denied");
            context.client().gui.hud.setOverlayMessage(message, false);
        }));
        ClientPlayNetworking.registerGlobalReceiver(AccessSyncPayload.TYPE, (payload, context) ->
                context.client().execute(() -> RoadMapScreen26.applyAccessAllowed(payload.allowed())));
    }

    public static void requestSnapshot(int requestSeq, Identifier dimensionId, int minX, int minZ, int maxX, int maxZ) {
        ClientPlayNetworking.send(new RequestRectPayload(requestSeq, dimensionId, minX, minZ, maxX, maxZ));
    }
    public static void requestTeleport(int x, int y, int z) { ClientPlayNetworking.send(new TeleportRequestPayload(x, y, z)); }
    public static void requestManualConnect(int ax, int az, int bx, int bz) { ClientPlayNetworking.send(new ManualConnectPayload(ax, az, bx, bz)); }
    public static void syncMapAccess(ServerPlayer player) {
        if (player != null) ServerPlayNetworking.send(player, new AccessSyncPayload(MapAccessService.canOpenMap(player)));
    }

    private record RequestRectPayload(int requestSeq, Identifier dimensionId, int minX, int minZ, int maxX, int maxZ) implements CustomPacketPayload {
        private static final Type<RequestRectPayload> TYPE = new Type<>(Identifier.fromNamespaceAndPath("roadweaver", "map_request_rect"));
        private static final StreamCodec<RegistryFriendlyByteBuf, RequestRectPayload> CODEC = CustomPacketPayload.codec(
                (p, b) -> { b.writeVarInt(p.requestSeq); b.writeIdentifier(p.dimensionId); b.writeVarInt(p.minX); b.writeVarInt(p.minZ); b.writeVarInt(p.maxX); b.writeVarInt(p.maxZ); },
                b -> new RequestRectPayload(b.readVarInt(), b.readIdentifier(), b.readVarInt(), b.readVarInt(), b.readVarInt(), b.readVarInt()));
        @Override public Type<? extends CustomPacketPayload> type() { return TYPE; }
    }
    private record SnapshotPayload(int requestSeq, Identifier dimensionId, MapSnapshot snapshot) implements CustomPacketPayload {
        private static final Type<SnapshotPayload> TYPE = new Type<>(Identifier.fromNamespaceAndPath("roadweaver", "map_snapshot"));
        private static final StreamCodec<RegistryFriendlyByteBuf, SnapshotPayload> CODEC = CustomPacketPayload.codec(
                (p, b) -> { b.writeVarInt(p.requestSeq); b.writeIdentifier(p.dimensionId); MapSnapshotCodec.write(b, p.snapshot); },
                b -> new SnapshotPayload(b.readVarInt(), b.readIdentifier(), MapSnapshotCodec.read(b)));
        @Override public Type<? extends CustomPacketPayload> type() { return TYPE; }
    }
    private record TeleportRequestPayload(int x, int y, int z) implements CustomPacketPayload {
        private static final Type<TeleportRequestPayload> TYPE = new Type<>(Identifier.fromNamespaceAndPath("roadweaver", "map_teleport"));
        private static final StreamCodec<RegistryFriendlyByteBuf, TeleportRequestPayload> CODEC = CustomPacketPayload.codec(
                (p, b) -> { b.writeVarInt(p.x); b.writeVarInt(p.y); b.writeVarInt(p.z); },
                b -> new TeleportRequestPayload(b.readVarInt(), b.readVarInt(), b.readVarInt()));
        @Override public Type<? extends CustomPacketPayload> type() { return TYPE; }
    }
    private record TeleportAckPayload(boolean ok, int x, int y, int z) implements CustomPacketPayload {
        private static final Type<TeleportAckPayload> TYPE = new Type<>(Identifier.fromNamespaceAndPath("roadweaver", "map_teleport_ack"));
        private static final StreamCodec<RegistryFriendlyByteBuf, TeleportAckPayload> CODEC = CustomPacketPayload.codec(
                (p, b) -> { b.writeBoolean(p.ok); b.writeVarInt(p.x); b.writeVarInt(p.y); b.writeVarInt(p.z); },
                b -> new TeleportAckPayload(b.readBoolean(), b.readVarInt(), b.readVarInt(), b.readVarInt()));
        @Override public Type<? extends CustomPacketPayload> type() { return TYPE; }
    }
    private record ManualConnectPayload(int ax, int az, int bx, int bz) implements CustomPacketPayload {
        private static final Type<ManualConnectPayload> TYPE = new Type<>(Identifier.fromNamespaceAndPath("roadweaver", "map_manual_connect"));
        private static final StreamCodec<RegistryFriendlyByteBuf, ManualConnectPayload> CODEC = CustomPacketPayload.codec(
                (p, b) -> { b.writeVarInt(p.ax); b.writeVarInt(p.az); b.writeVarInt(p.bx); b.writeVarInt(p.bz); },
                b -> new ManualConnectPayload(b.readVarInt(), b.readVarInt(), b.readVarInt(), b.readVarInt()));
        @Override public Type<? extends CustomPacketPayload> type() { return TYPE; }
    }
    private record AccessSyncPayload(boolean allowed) implements CustomPacketPayload {
        private static final Type<AccessSyncPayload> TYPE = new Type<>(Identifier.fromNamespaceAndPath("roadweaver", "map_access_sync"));
        private static final StreamCodec<RegistryFriendlyByteBuf, AccessSyncPayload> CODEC = CustomPacketPayload.codec(
                (p, b) -> b.writeBoolean(p.allowed), b -> new AccessSyncPayload(b.readBoolean()));
        @Override public Type<? extends CustomPacketPayload> type() { return TYPE; }
    }
}

File: fabric/src/main/kotlin/net/shiroha233/roadweaver/network/fabric/MapNetworkFabric.kt
package net.shiroha233.roadweaver.network.fabric

import net.fabricmc.fabric.api.client.networking.v1.ClientPlayNetworking
import net.fabricmc.fabric.api.networking.v1.PacketByteBufs
import net.fabricmc.fabric.api.networking.v1.ServerPlayNetworking
import net.minecraft.core.BlockPos
import net.minecraft.network.chat.Component
import net.minecraft.resources.ResourceLocation
import net.minecraft.world.level.levelgen.Heightmap
import net.shiroha233.roadweaver.RoadWeaver
import net.shiroha233.roadweaver.client.map.RoadMapScreen
import net.shiroha233.roadweaver.client.map.data.MapDataCollector
import net.shiroha233.roadweaver.helpers.Records
import net.shiroha233.roadweaver.network.MapSnapshotCodec
import net.shiroha233.roadweaver.persistence.WorldDataProvider
import net.shiroha233.roadweaver.util.ComputeService
import java.util.concurrent.CompletableFuture

object MapNetworkFabric {
    val REQ_RECT_ID: ResourceLocation = ResourceLocation(RoadWeaver.MOD_ID, "map_request_rect")
    val SNAP_ID: ResourceLocation = ResourceLocation(RoadWeaver.MOD_ID, "map_snapshot")
    val TP_REQ_ID: ResourceLocation = ResourceLocation(RoadWeaver.MOD_ID, "map_teleport")
    val TP_ACK_ID: ResourceLocation = ResourceLocation(RoadWeaver.MOD_ID, "map_teleport_ack")
    val MAN_REQ_ID: ResourceLocation = ResourceLocation(RoadWeaver.MOD_ID, "map_manual_connect")

    @JvmStatic
    fun register() {
        // Nothing needed here for 1.20.1 style, Handlers registered during init
    }

    @JvmStatic
    fun registerServerReceivers() {
        ServerPlayNetworking.registerGlobalReceiver(REQ_RECT_ID) { server, player, handler, buf, responseSender ->
            val minX = buf.readVarInt()
            val minZ = buf.readVarInt()
            val maxX = buf.readVarInt()
            val maxZ = buf.readVarInt()
            
            val cx = kotlin.math.round(player.x).toInt()
            val cz = kotlin.math.round(player.z).toInt()

            val radiusChunks: Int = try {
                val cfg = net.shiroha233.roadweaver.config.ConfigService.get()
                if (cfg.dynamicPlanEnabled()) cfg.dynamicPlanRadiusChunks() else cfg.initialPlanRadiusChunks()
            } catch (_: Throwable) { 256 }
            val radiusBlocks = kotlin.math.max(1, radiusChunks) * 16

            val level = player.serverLevel()
            CompletableFuture.supplyAsync({
                MapDataCollector.build(level, minX, minZ, maxX, maxZ, cx, cz, radiusBlocks)
            }, ComputeService.executor()).thenAccept { snapshot ->
                server.execute {
                    val out = PacketByteBufs.create()
                    MapSnapshotCodec.write(out, snapshot)
                    ServerPlayNetworking.send(player, SNAP_ID, out)
                }
            }
        }

        ServerPlayNetworking.registerGlobalReceiver(TP_REQ_ID) { server, player, handler, buf, responseSender ->
            val x = buf.readVarInt()
            val y = buf.readVarInt()
            val z = buf.readVarInt()
            
            server.execute {
                val allowed = player.isCreative || player.hasPermissions(2)
                if (!allowed) {
                    val ack = PacketByteBufs.create()
                    ack.writeBoolean(false)
                    ack.writeVarInt(0); ack.writeVarInt(0); ack.writeVarInt(0)
                    ServerPlayNetworking.send(player, TP_ACK_ID, ack)
                    return@execute
                }

                val level = player.serverLevel()
                level.getChunk(x shr 4, z shr 4)
                var ty = level.getHeight(Heightmap.Types.MOTION_BLOCKING_NO_LEAVES, x, z)
                ty = if (ty <= level.minBuildHeight) level.seaLevel + 1 else ty + 1

                player.teleportTo(level, x + 0.5, ty.toDouble(), z + 0.5, player.yRot, player.xRot)
                
                val ack = PacketByteBufs.create()
                ack.writeBoolean(true)
                ack.writeVarInt(x); ack.writeVarInt(ty); ack.writeVarInt(z)
                ServerPlayNetworking.send(player, TP_ACK_ID, ack)
            }
        }

        ServerPlayNetworking.registerGlobalReceiver(MAN_REQ_ID) { server, player, handler, buf, responseSender ->
            val ax = buf.readVarInt()
            val az = buf.readVarInt()
            val bx = buf.readVarInt()
            val bz = buf.readVarInt()
            
            server.execute {
                val level = player.serverLevel()
                val provider = WorldDataProvider.getInstance()
                val origin = provider.getStructureConnections(level)
                val list = if (origin != null) ArrayList(origin) else ArrayList()

                val a = BlockPos(ax, 0, az)
                val b = BlockPos(bx, 0, bz)

                var exists = false
                for (c in list) {
                    val f = c.from
                    val t = c.to
                    val sameAB = f.x == a.x && f.z == a.z && t.x == b.x && t.z == b.z
                    val sameBA = f.x == b.x && f.z == b.z && t.x == a.x && t.z == a.z
                    if (sameAB || sameBA) {
                        exists = true
                        break
                    }
                }

                if (!exists) {
                    list.add(Records.StructureConnection(a, b, Records.ConnectionStatus.PLANNED))
                    provider.setStructureConnections(level, list)
                }
            }
        }
    }

    @JvmStatic
    fun registerClientReceivers() {
        ClientPlayNetworking.registerGlobalReceiver(SNAP_ID) { client, handler, buf, responseSender ->
            val snapshot = MapSnapshotCodec.read(buf)
            client.execute {
                val current = client.screen
                if (current is RoadMapScreen) {
                    current.onMapSnapshotReceived(snapshot, 0)
                }
            }
        }

        ClientPlayNetworking.registerGlobalReceiver(TP_ACK_ID) { client, handler, buf, responseSender ->
            val ok = buf.readBoolean()
            val fx = buf.readVarInt()
            val fy = buf.readVarInt()
            val fz = buf.readVarInt()
            client.execute {
                val p = client.player ?: return@execute
                if (ok) p.displayClientMessage(Component.translatable("gui.roadweaver.map.teleport.success_pos", fx, fy, fz), true)
                else p.displayClientMessage(Component.translatable("gui.roadweaver.map.teleport.denied"), true)
            }
        }
    }

    @JvmStatic
    fun requestSnapshot(minX: Int, minZ: Int, maxX: Int, maxZ: Int) {
        val buf = PacketByteBufs.create()
        buf.writeVarInt(minX)
        buf.writeVarInt(minZ)
        buf.writeVarInt(maxX)
        buf.writeVarInt(maxZ)
        ClientPlayNetworking.send(REQ_RECT_ID, buf)
    }

    @JvmStatic
    fun requestTeleport(x: Int, y: Int, z: Int) {
        val buf = PacketByteBufs.create()
        buf.writeVarInt(x)
        buf.writeVarInt(y)
        buf.writeVarInt(z)
        ClientPlayNetworking.send(TP_REQ_ID, buf)
    }

    @JvmStatic
    fun requestManualConnect(ax: Int, az: Int, bx: Int, bz: Int) {
        val buf = PacketByteBufs.create()
        buf.writeVarInt(ax)
        buf.writeVarInt(az)
        buf.writeVarInt(bx)
        buf.writeVarInt(bz)
        ClientPlayNetworking.send(MAN_REQ_ID, buf)
    }
}
